### ISSUE TITLE
chart.exporting.dateFormat = "h a"; not doing anything in csv export

### ISSUE THREAD

**Issue by morrow95** [amcharts/amcharts4#1372]
Chart is working fine, but I am having trouble formatting a date column in the export.  The 'Timestamp' column has values in the format of XX representing the hours of the day... like 00, 01, 02, ... 23.  In the actual chart display I can format these values to with "h a" so they show like 10 AM, 6 PM, etc.  I am having trouble applying the same formatting in the exports.

```
//makes sure the exports have the correct date format
chart.exporting.dateFields.push("Timestamp");
chart.exporting.dateFormat = "h a";
```

Fiddle is here [https://jsfiddle.net/57jvL1a4/](https://jsfiddle.net/57jvL1a4/) with an example and setup for exports... just click csv to get the csv export.

**Comment by morrow95** [comment]
As a workaround for the time being I can use the following to format these properly in the exports : 

```
//need to manually format the hours column because we can't get the regular date/time to work
chart.exporting.adapter.add("data", function(data) {
	for (var i = 0; i < data.data.length; i++) {
		var dd = 'AM';
		var h = hh = data.data[i].Timestamp;
		
		//handle 12PM and above
		if (h >= 12) {
			h = hh - 12;
			dd = "PM";
		}
		//handle 0 AM
		if (h == 0) {
			h = 12;
		} 
		//save it
		data.data[i].Timestamp = parseInt(h, 10) + ' ' + dd;
  }
  return data;
});

//these should work, but they don't so we handle it above manually
//makes sure the exports have the correct date format
//chart.exporting.dateFields.push("Timestamp");
//chart.exporting.dateFormat = "h a";
```

but I would still like to know why the 'built-in' formatting does not work in this scenario.

**Comment by martynasma** [comment]
Well, the thing is that `dateFormat` works for Date objects only. You have strings like `"00"` as `Timestamp`, hence date formatter not running on it.

I suppose the adapter approach is the easiest way to go about it is using adapters, like you currently have.

**Comment by morrow95** [comment]
The only place in the docs you can format time in any way is within the date formatting options you offer.  If you look at the fiddle I showed I am able to correctly format 2 digit hours using 'h a' and date formatting everywhere, but the exporting.  The chart itself formats and shows as it should... it is only the exporting.  For instance : 

```
chart.dateFormatter.inputDateFormat = "HH";

dateAxis.baseInterval = {
	"timeUnit": "hour",
	"count": 1
};
dateAxis.dateFormats.setKey("hour", "h a");
dateAxis.periodChangeDateFormats.setKey("hour", "h a");
```

**Comment by martynasma** [comment]
We'll explore the possibility of Export using chart's `inputDateFormat` to parse string-based dates. Thanks.

Will keep you posted.

**Comment by martynasma** [comment]
Fixed in [4.5.6](https://github.com/amcharts/amcharts4/releases/tag/4.5.6).

## [4.5.6] - 2019-07-16

### Added
- New property on `Annotation` plugin: `data`. Allows getting serialized annotation data object, as well as setting it back to load annotations.

### Changed
- Annotation plugin will not create full Export menu if it wasn't explicitly created before.
- Tweaked CSS for `ExportMenu` to protect from icons getting small becuase of the page-wide CSS.
- When exporting data on chart with `exporting.dateFormat` set, the date/time fields in data will be formatted according to the set format.

### Fixed
- `XYChartScrollbar` with series attached to different `ValueAxis` were using same axis scale in the scrollbar.
- Using `wrap = true` on horizontal axis labels was resulting weird behavior when zooming in some cases.
- Export was not prioritizing menu item's specific options over generic format options.
- Auto-wrapped text was not retaining square bracket escaping (double bracket) for each subsequent line.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.

**Comment by morrow95** [comment]
@martynasma 

Looks like there was some collateral damage done with this change.  I am showing 'Invalid date' as the column header.  You can see this by using the original fiddle I posted in this thread, [https://jsfiddle.net/57jvL1a4/](https://jsfiddle.net/57jvL1a4/), and clicking on the csv export option.

In this case, the column headers should be output as Hour and Total due to : 

chart.exporting.dataFields = { Timestamp: 'Hour', Total: 'Total' };

If you click on a different export option, like JSON, it will output correctly.  Just a guess, but looks like you are processing the header somewhere when that should be left alone.  I just tested 4.5.6 and noticed this.  This did not happen with 4.4.10 which was the previous version I was using.

**Comment by martynasma** [comment]
Oh boy. Indeed. Fixing.

**Comment by martynasma** [comment]
Fixed in [4.5.7](https://github.com/amcharts/amcharts4/releases/tag/4.5.7).

## [4.5.7] - 2019-07-18

### Added
- New plugin: [Bullets](https://www.amcharts.com/docs/v4/tutorials/plugin-bullets/). It's a collection of configurable shapes (stars, pins, flags, etc.) that you can use as chart bullets, map markers, and anywhere else.
- New `Tooltip` property: `showInViewport` (default `false`). if set to `true` it will not hide tooltips that point to an off-viewport location. It will be glued to the closes viewport edge instead.

### Changed
- Tooltip for `MapPolygonSeries` now has `showInViewport = true` set by default.

### Fixed
- Exporting to CSV with `dateFormat` set was causing "Invalid date" to appear instead of date column names (since 4.5.6).
- When `MapPolygonSeries` had only one item set in its `include` array, it was causing map to come up as empty in some cases.
- Charts were not accounting for their container padding/borders when measuring their own size, causing bigger charts then necessary.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
